design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,312 @@
+use cast::starknet_commands::account::create::create as create_account
+(...)
+
+fn make_account(...) -> Result<InvokeResponse> {
+  let user = create_account(...);
+  let prefund_account = invoke(...);
+  deploy_account(...);
+}
+
+fn main() {
+  let user = make_account(...);
+  skip_if_done(|| {
+        dispatcher.increase_balance(...);
+    });
+  let get_balance = skip_if_done(|| {
+        dispatcher.get_balance(...);
+    });
+}  
+```
+
+### example state file
+The state file by default should be written to the current working directory, with a name `<script file name>.state.json`.
+Its schema could look like this:
+
+```json
+{
+  "create_account": {
+    "arguments": {
+      "name": "whatever",
+      (...)
+    },
+    "output": {
+      "address": "0x123",
+      "max_fee": "0x321",
+    },
+    "status": "accepted",
+    "timestamp": (...),
+  },
+  "prefund_account": {
+    "arguments": {
+      ...
+    },
+    "output": {
+      "error": "..."
+    }
+  },
+  "status": "rejected",
+  "timestamp": (...),
+}
+```
+
+Having this, we could allow users to `--resume`, `--retry` or `--verify` transactions if something unexpected happens,
+some transactions were rejected because of faulty script, too low user eth balance etc.
+
+If we were to implement dry run mode that runs by default, the state should only be written to a file when we pass `--broadcast`
+flag; otherwise the output should be directed to stdout.
+
+### Miscellaneous
+
+There would be a new subcommand `script` that would be invoked with required parameter - a path to script file, with
+`*.s.cairo` extension. Proposed interface:
+
+- `--gas-multiplier` - relative percentage by which to multiply add gas estimates (perhaps passed directly to sncast)
+- `--boradcast` - boradcasts the transactions
+- `--no-state-file` - do not write to state file
+- `--to-state-file` - specify path to custom state file
+- `--log-file` - output logs to a file
+- `--silent` - do not output logs to stdout
+- `--slow` - makes sure a transaction is sent, only after its previous one has been confirmed and succeeded (possible reuse of `--wait` flag)
+- `--verify` - find a matching broadcast in state file, and try to verify transactions
+- `--delay` - optional timeout to apply in between attempts in seconds (perhaps passed directly to sncast)
+- `--retries` - number of attempts for retrying (perhaps passed directly to sncast)
+
+If it makes sense, some of the flags could be included in Scarb.toml config file.
+
+## Proposed steps/tasks
+
+MVP:
+  - allow for writing scripts using dispatchers and imported cast subcommand functions
+    - no idempotency required at this stage
+    - all cast subcommand functions return every information necessary (tx hashes, addresses etc) as importable structs
+    - no cast subcommand requires CastConfig directly (eg as parameter)
+    - it is possible to create and deploy an account, create and deploy a contract
```

Comment:
I assume importing a contract would be possible too, right?

Reply:
yes: `allow for writing scripts using dispatchers`

Reply:
Sorry, I meant the `account`

Reply:
Yes, this would be doable using `get_account` function (as in example script)